tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/OboTests.cs

```diff
@@ -51,7 +52,279 @@ public void TestInitialize()
+            Assert.AreEqual(1, cca.UserTokenCacheInternal.Accessor.GetAllAccessTokens().Count);
+            Assert.AreEqual(1, cca.UserTokenCacheInternal.Accessor.GetAllRefreshTokens().Count);
+
+            result = await cca.AcquireTokenOnBehalfOf(s_scopes, userAssertion).ExecuteAsync().ConfigureAwait(false);
+
+            // Cache has 2 partitions (user-provided key, assertion) with 1 token each
+            Assert.IsNotNull(result);
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+            Assert.AreEqual(2, cca.UserTokenCacheInternal.Accessor.GetAllAccessTokens().Count);
+            Assert.AreEqual(1, cca.UserTokenCacheInternal.Accessor.GetAllRefreshTokens().Count);
+
+            // Returns long-running token
+            result = await cca.AcquireTokenInLongRunningProcess(s_scopes, oboCacheKey).ExecuteAsync().ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+
+            // Returns normal token
+            result = await cca.AcquireTokenOnBehalfOf(s_scopes, userAssertion).ExecuteAsync().ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+        }
+
+        /// <summary>
+        /// Tests the behavior when calling both, long-running and normal OBO methods.
+        /// Both methods should return the same tokens, since the cache key is the same.
+        /// Should be the same partition: by assertion hash.
+        /// </summary>
+        [TestMethod]
+        public async Task AcquireTokenByObo_LongRunningThenNormalObo_WithTheSameKey_TestAsync()
+        {
+            var user1 = (await LabUserHelper.GetSpecificUserAsync("idlab1@msidlab4.onmicrosoft.com").ConfigureAwait(false)).User;
+            var pca = PublicClientApplicationBuilder
+                .Create(PublicClientID)
+                .WithAuthority(AadAuthorityAudience.AzureAdMultipleOrgs)
+                .Build();
+
+            var userAuthResult = await pca
+                .AcquireTokenByUsernamePassword(s_oboServiceScope, user1.Upn, new NetworkCredential("", user1.GetOrFetchPassword()).SecurePassword)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            var cca = BuildCCA(userAuthResult.TenantId);
+
+            string oboCacheKey = null;
+            UserAssertion userAssertion = new UserAssertion(userAuthResult.AccessToken);
+
+            // InitiateLR - Empty cache - AT via OBO flow (new AT, RT cached)
+            var result = await cca.InitiateLongRunningProcessInWebApi(s_scopes, userAuthResult.AccessToken, ref oboCacheKey)
+                                    .ExecuteAsync().ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
```

Comment:
nit: add an assertion that oboCacheKey is now `hash(userAssertion)`

Reply:
Added.